# Logistic Regression case study 

CSV file which tells which of the users purchased/not purchased a particular product.
I have used a logistic regression model to predict the purchase based on certain variables like gender, Age and Estimated Salary 

## Importing libraries 

In [1]:
import numpy as np 
import pandas as pd 
import matplotlib.pyplot as plt 

## Importing dataset

https://www.kaggle.com/rakeshrau/social-network-ads

In [2]:
dataset = pd.read_csv('C:/Users/binayak/Videos/Machine Learning/Classification/Logistic Regression/Social_Network_Ads.csv')

In [3]:
dataset

User ID  Gender  Age  EstimatedSalary  Purchased
0    15624510    Male   19            19000          0
1    15810944    Male   35            20000          0
2    15668575  Female   26            43000          0
3    15603246  Female   27            57000          0
4    15804002    Male   19            76000          0
..        ...     ...  ...              ...        ...
395  15691863  Female   46            41000          1
396  15706071    Male   51            23000          1
397  15654296  Female   50            20000          1
398  15755018    Male   36            33000          0
399  15594041  Female   49            36000          1

[400 rows x 5 columns]

In [4]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 400 entries, 0 to 399
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   User ID          400 non-null    int64 
 1   Gender           400 non-null    object
 2   Age              400 non-null    int64 
 3   EstimatedSalary  400 non-null    int64 
 4   Purchased        400 non-null    int64 
dtypes: int64(4), object(1)
memory usage: 14.1+ KB


In [5]:
dataset.describe()

User ID         Age  EstimatedSalary   Purchased
count  4.000000e+02  400.000000       400.000000  400.000000
mean   1.569154e+07   37.655000     69742.500000    0.357500
std    7.165832e+04   10.482877     34096.960282    0.479864
min    1.556669e+07   18.000000     15000.000000    0.000000
25%    1.562676e+07   29.750000     43000.000000    0.000000
50%    1.569434e+07   37.000000     70000.000000    0.000000
75%    1.575036e+07   46.000000     88000.000000    1.000000
max    1.581524e+07   60.000000    150000.000000    1.000000

## Separate the dependent variable from independent 

In [6]:
X = dataset.iloc[:,1:-1].values
y = dataset.iloc[:,-1].values

In [7]:
X

array([['Male', 19, 19000],
       ['Male', 35, 20000],
       ['Female', 26, 43000],
       ...,
       ['Female', 50, 20000],
       ['Male', 36, 33000],
       ['Female', 49, 36000]], dtype=object)

In [8]:
y

array([0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1,
       0, 1, 1, 1, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 0, 1, 1, 0,
       1, 1, 0, 1, 0, 1, 0, 1, 0, 0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 1, 1, 0,
       1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1,
       0, 1, 0, 1, 1, 1, 1, 0, 0, 0, 1, 1, 0, 1, 1,

## Categorical variables : Encoded

In [9]:
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer

ct = ColumnTransformer(transformers=[('encoder',OneHotEncoder(),[0])],
                       remainder='passthrough')

In [10]:
X= np.array(ct.fit_transform(X))

In [11]:
X

array([[0.0, 1.0, 19, 19000],
       [0.0, 1.0, 35, 20000],
       [1.0, 0.0, 26, 43000],
       ...,
       [1.0, 0.0, 50, 20000],
       [0.0, 1.0, 36, 33000],
       [1.0, 0.0, 49, 36000]], dtype=object)

## Split into Training and Test set

In [12]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=1/4)

In [13]:
print('The shape of X_train: {}; y_train: {}; X_test:{}; y_test:{}'.
      format(X_train.shape,y_train.shape,X_test.shape,y_test.shape))

The shape of X_train: (300, 4); y_train: (300,); X_test:(100, 4); y_test:(100,)


## Feature Scaling 

In [14]:
from sklearn.preprocessing import StandardScaler
ss = StandardScaler()
X_train[:,2:] = ss.fit_transform(X_train[:,2:])
X_test[:,2:] = ss.transform(X_test[:,2:])

In [15]:
X_train[:10,:]

array([[0.0, 1.0, 0.23292084944636707, -0.3447292668153169],
       [1.0, 0.0, -1.0980554331043004, 0.3411507623155039],
       [1.0, 0.0, -0.1473580884252522, 0.19204640815662982],
       [1.0, 0.0, -0.43256729182896664, 0.01312118316598092],
       [0.0, 1.0, 0.23292084944636707, -0.642937975133065],
       [1.0, 0.0, -1.0029856986363956, -1.5375641000863096],
       [1.0, 0.0, -1.6684738399117294, -1.3586388750956606],
       [0.0, 1.0, 1.0885484596575103, 0.6095385998014773],
       [0.0, 1.0, -1.7635435743796342, 0.04294205399775573],
       [0.0, 1.0, -0.242427822893157, -0.3149083959835421]], dtype=object)

In [16]:
X_test[:10,:]

array([[0.0, 1.0, -0.05228835395734738, -0.493833620974191],
       [1.0, 0.0, 1.6589668664649393, 1.8321943039042448],
       [1.0, 0.0, 0.5181300528500815, 1.295418628932298],
       [1.0, 0.0, 1.2786879285933201, 1.951477787231344],
       [0.0, 1.0, 0.13785111497846225, 1.951477787231344],
       [0.0, 1.0, -0.1473580884252522, -1.0604301667779126],
       [0.0, 1.0, 0.7082695217858912, -1.3884597459274355],
       [0.0, 1.0, -1.0980554331043004, 0.3411507623155039],
       [1.0, 0.0, -1.0029856986363956, -0.43419187931064135],
       [1.0, 0.0, 0.13785111497846225, 0.07276292482953055]], dtype=object)

## Training the Logistic model on training set

In [17]:
from sklearn.linear_model import LogisticRegression
logreg = LogisticRegression()
logreg.fit(X_train,y_train)

LogisticRegression()

## Predict purchase in test set

In [18]:
y_pred = logreg.predict(X_test)
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.reshape(len(y_test),1)),
                    axis=1))

[[0 0]
 [1 1]
 [1 1]
 [1 0]
 [1 1]
 [0 0]
 [0 1]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [1 1]
 [1 1]
 [0 1]
 [0 0]
 [0 0]
 [1 0]
 [0 0]
 [1 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 1]
 [1 1]
 [0 0]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [1 0]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [0 1]
 [0 1]
 [0 0]
 [0 0]
 [0 0]
 [1 0]
 [1 1]
 [0 1]
 [0 0]
 [1 1]
 [0 0]
 [1 0]
 [0 0]
 [0 0]
 [1 1]
 [1 1]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [1 1]
 [0 0]
 [0 0]
 [1 1]
 [1 1]
 [1 1]
 [1 1]
 [0 0]
 [1 1]
 [0 0]
 [1 0]
 [1 1]
 [0 0]
 [1 1]
 [1 1]
 [0 0]
 [0 0]
 [0 1]
 [1 1]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [0 0]
 [1 1]
 [0 0]
 [0 0]]


## Making a Confusion Matrix & Accuracy score

In [19]:
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test,y_pred)
print(cm)

from sklearn.metrics import accuracy_score
a_score = accuracy_score(y_test,y_pred)
print(a_score)

[[57  7]
 [ 7 29]]
0.86


57 : correct prediction of class 0 

29 : correct prediction of class 1 

Hence 86% are correct (since test set has 100 observations)

Thus,

7 : incorrect prediction of class 1

7 : incorrect prediction of class 0
